File: OS Project .py
import random

class PageReplacement:
    def __init__(self):
        self.size = 0
        self.frame_size = 0
        self.array = []

    def input_data(self, algo_type):
        self.size = int(input("Enter Size of Data: "))
        if algo_type == 1:
            max_range = int(input("Enter Maximum Range of Reference String: "))
            self.array = [random.randint(0, max_range) for _ in range(self.size)]
            print("Reference String is: ", self.array)
        elif algo_type == 2:
            self.array = list(map(int, input("Enter Reference String (space-separated): ").split()))
        
        self.frame_size = int(input("Enter Frame Size: "))
        print("\nAlgorithm\tPage Faults\tHit Count\tPage Fault Ratio\tHit Ratio")

class FIFO:
    def __init__(self, size, frame_size, array):
        self.size = size
        self.frame_size = frame_size
        self.array = array
        self.queue = []
        self.page_faults = 0
        self.hits = 0

    def run(self):
        for page in self.array:
            if page not in self.queue:
                if len(self.queue) < self.frame_size:
                    self.queue.append(page)
                else:
                    self.queue.pop(0)
                    self.queue.append(page)
                self.page_faults += 1
            else:
                self.hits += 1
        
        self.print_results("FIFO")

    def print_results(self, algo_name):
        fault_ratio = (self.page_faults / self.size) * 100
        hit_ratio = (self.hits / self.size) * 100
        print(f"{algo_name}\t\t{self.page_faults}\t\t{self.hits}\t\t{fault_ratio:.2f}%\t\t{hit_ratio:.2f}%")

class LRU:
    def __init__(self, size, frame_size, array):
        self.size = size
        self.frame_size = frame_size
        self.array = array
        self.queue = []
        self.page_faults = 0
        self.hits = 0

    def run(self):
        for page in self.array:
            if page not in self.queue:
                if len(self.queue) < self.frame_size:
                    self.queue.append(page)
                else:
                    # Evict the least recently used (front of queue)
                    self.queue.pop(0)
                    self.queue.append(page)
                self.page_faults += 1
            else:
                # Move accessed page to the most recently used position
                self.queue.remove(page)
                self.queue.append(page)
                self.hits += 1
        
        self.print_results("LRU")

    def print_results(self, algo_name):
        fault_ratio = (self.page_faults / self.size) * 100
        hit_ratio = (self.hits / self.size) * 100
        print(f"{algo_name}\t\t{self.page_faults}\t\t{self.hits}\t\t{fault_ratio:.2f}%\t\t{hit_ratio:.2f}%")

class SCA:
    def __init__(self, size, frame_size, array):
        self.size = size
        self.frame_size = frame_size
        self.array = array
        self.stack = []
        self.bits = []
        self.page_faults = 0
        self.hits = 0

    def run(self):
        for page in self.array:
            if page not in self.stack:
                if len(self.stack) < self.frame_size:
                    self.stack.append(page)
                    self.bits.append(0)
                else:
                    while True:
                        if self.bits[0] == 0:
                            self.stack.pop(0)
                            self.bits.pop(0)
                            self.stack.append(page)
                            self.bits.append(0)
                            break
                        else:
                            self.bits.pop(0)
                            self.bits.append(0)
                            self.stack.append(self.stack.pop(0))
                self.page_faults += 1
            else:
                idx = self.stack.index(page)
                self.bits[idx] = 1
                self.hits += 1

        self.print_results("SCA")

    def print_results(self, algo_name):
        fault_ratio = (self.page_faults / self.size) * 100
        hit_ratio = (self.hits / self.size) * 100
        print(f"{algo_name}\t\t{self.page_faults}\t\t{self.hits}\t\t{fault_ratio:.2f}%\t\t{hit_ratio:.2f}%")

def menu():
    print("\n1-> Enter Data Automatically")
    print("2-> Enter Data Manually")
    print("3-> Exit")

    choice = int(input("\nEnter your choice: "))
    if choice in [1, 2]:
        page_replacement = PageReplacement()
        page_replacement.input_data(choice)

        fifo = FIFO(page_replacement.size, page_replacement.frame_size, page_replacement.array)
        lru = LRU(page_replacement.size, page_replacement.frame_size, page_replacement.array)
        sca = SCA(page_replacement.size, page_replacement.frame_size, page_replacement.array)

        fifo.run()
        lru.run()
        sca.run()
    elif choice == 3:
        print("\nThank you for using the system!")
        exit()
    else:
        print("Invalid choice! Try again.")
        menu()

if __name__ == "__main__":
    menu()

File: OS project.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
using namespace std;

int size;
int *Array;
int frame_size;
int max_range;

void Input(int Algo_type) {
    cout << "Enter Size of Data: ";
    cin >> size;
    Array = new int[size];
    
    if (Algo_type == 1) {
        cout << "Enter Maximum Range of Reference String: ";
        cin >> max_range;
        srand(time(0));
        cout << "Reference String is: ";
        for (int i = 0; i < size; i++) {
            Array[i] = rand() % (max_range + 1);
            cout << Array[i] << " ";
        }
        cout << endl;
    } else if (Algo_type == 2) {
        cout << "Enter Data You Want (Reference String): ";
        for (int i = 0; i < size; i++) {
            cin >> Array[i];
        }
    }

    cout << "Enter Frame Size: ";
    cin >> frame_size;
    cout << "\nAlgorithm\tPageFaults\tHitCount\tPage_Fault-Ratio\tHit-Ratio\n";
}

class BitArray {
public:
    int data;
    bool bit;
    BitArray(int value = -1, bool bol = false) : data(value), bit(bol) {}
};

class Node {
public:
    int data;
    Node *next;
    Node(int value = -1, Node *nood = nullptr) : data(value), next(nood) {}
};

class Queue {
    Node *head;
public:
    Queue() : head(nullptr) {}

    void Enqueue(int value) {
        Node *newNode = new Node(value);
        if (!head) {
            head = newNode;
        } else {
            Node *temp = head;
            while (temp->next) {
                temp = temp->next;
            }
            temp->next = newNode;
        }
    }

    void Dequeue() {
        if (!head) return;
        Node *temp = head;
        head = head->next;
        delete temp;
    }

    bool Check(int value) {
        Node *temp = head;
        while (temp) {
            if (temp->data == value) return true;
            temp = temp->next;
        }
        return false;
    }

    int Size() {
        int count = 0;
        Node *temp = head;
        while (temp) {
            count++;
            temp = temp->next;
        }
        return count;
    }

    bool IsEmpty() {
        return head == nullptr;
    }

    void Remove(int value) {
        if (!head) return;
        if (head->data == value) {
            Dequeue();
            return;
        }
        Node *temp = head;
        while (temp->next && temp->next->data != value) {
            temp = temp->next;
        }
        if (temp->next) {
            Node *toDelete = temp->next;
            temp->next = toDelete->next;
            delete toDelete;
        }
    }

    void Traverse() {
        Node *temp = head;
        while (temp) {
            cout << temp->data << " ";
            temp = temp->next;
        }
        cout << endl;
    }

    ~Queue() {
        while (!IsEmpty()) {
            Dequeue();
        }
    }
};

// Add similar corrections for `Stack`, `FIFO`, `LRU`, and `SCA` classes.

void End() {
    cout << "\n\nThank you for using this system!\n";
    exit(0);
}

int choice(int x) {
    int choise;
    cout << "\nEnter your choice: ";
    cin >> choise;
    while (choise < 1 || choise > x) {
        cout << "Invalid choice! Try again: ";
        cin >> choise;
    }
    return choise;
}

int Menu() {
    cout << "\n1-> Enter Data Automatically\n";
    cout << "2-> Enter Data Manually\n";
    cout << "3-> Exit\n";
    int x = choice(3);
    return x;
}

int main() {
    int x = Menu();
    if (x == 1 || x == 2) {
        Input(x);
        // Call algorithms here (FIFO, LRU, SCA).
        delete[] Array;
        main(); // Recursive for testing, replace with loop in production.
    } else {
        End();
    }
}
